pna.p4

```diff
@@ -665,19 +665,75 @@ extern void mirror_packet(MirrorSlotId_t mirror_slot_id,
 // probably be a parameter that specifies the new entry's initial
 // expire_time_profile_id.
 
-extern bool add_entry<T>(string action_name,
-                         in T action_params,
-                         in ExpireTimeProfileId_t expire_time_profile_id);
-
-// If we support the following variant of add_entry, with no
-// expire_time_profile_id parameter, then we need to define its
-// behavior in terms of what the added entry's value of
-// expire_time_profile_id is.  One option would be to define the
-// behavior to be the same as the 3-parameter version of add_entry
-// above, with expire_time_profile_id equal to 0.
-
-//extern bool add_entry<T>(string action_name,
-//                         in T action_params);
+// The bit width of this type is allowed to be different for different
+// target devices.  It must be at least a 1-bit wide type.
+
+typedef bit<1> AddEntryErrorStatus_t;
+
+const AddEntryErrorStatus_t ADD_ENTRY_SUCCESS = 0;
+const AddEntryErrorStatus_t ADD_ENTRY_NOT_DONE = 1;
+
+// Targets may define target-specific non-0 constants of type
+// AddEntryErrorStatus_t if they wish.
+
+// add_entry() must be called within the body of a P4 action that is
+// the default_action of a table with the table property `add_on_miss`
+// equal to true.
+//
+// The key of the entry added will always be the same as the key that
+// was just looked up in the table, and experienced a miss.
+//
+// `action_name` is the name of an action of this table.  It must be
+// possible for this action to be the action of an entry added to the
+// table, i.e. it is an error if the action has the annotation
+// `@defaultonly`.
```

Comment:
Should we not say "default_action" table property versus annotation "@deafult_only" ?

Reply:
I tried to be very precise about what I meant there, and I think what I wrote is what I meant, but see if the explanation below clarifies anything:

Some actions can be both hit actions and miss actions for a table, i.e. if they have neither a `@tableonly` nor a `@defaultonly` annotation them.  According to the text, it is OK for such an action to be the first parameter of an `add_entry` call.  That is true EVEN IF such an action is ALSO currently the default (miss) action of the table.  Why should we disallow that?  It might be unusual for an action to be a table's default action, and also installed as the action of one or more table entries, but the P4 language has allowed this since it was defined.

Reply:
Then may be description can also include both "@defaultonly" annotation or has  table property "default_action".

Reply:
What if a table is defined like this?
```
table t1 {
    key = { /* key fields not important for this example, so omitted */ }
    actions = {
        a1;
        a2;
    }
    default_action = a2;
}
```
I would argue that action `a2` is a perfectly acceptable argument to a call to `add_entry`, because it does not have a `@defaultonly` annotation.

The fact that it is the value of the `default_action` property does not stop the control plane from adding entries with action `a2`, so why should add_entry be disallowed from adding entries with action `a2`?

Reply:
From the description I understood as if default_action adds entry with a1 as action, then it should be accepted. However if default_action adds entry with "a2" as match-action for the new entry , then when "a2" happens to be default_action (@default_action annotation), then it should not be allowed. Let me re-read the description.

Reply:
default_action does not add any entries to the table.  Entries have a key, and an associated action with its action parameter values.  There is no key for a default_action.  When an .apply() call is performed on a table, the key value is searched to see if it matches any entries in the table.  If no installed entries match, then the table's current default_action is executed.  The default_action can be changed by the control plane any time it wants to, unless it is declared as `const default_action`.  Sorry if this is all well-known to you already, but it isn't clear to me where the confusion lies here.

Reply:
Would it be clearer if the restriction were documented in the following way instead?

`action_name` is the name of an action that must meet the following conditions:

+ it must be an action in the `actions` property for the table to which it is adding the entry
+ it must be allowed for the action to be an action associated with a table entry, i.e. it must not have the `@defaultonly` attribute
+ it must be an action that does not contain a call to `add_entry`

Reply:
The earlier description is clear. I cannot close/resolve comments. Pls consider any comments with "like" from my end are resolved from my end.

Reply:
Sorry, do you mind copying and pasting what you consider to be the "earlier description"?

Reply:
Never mind this question, as I believe you have answered it in this comment: https://github.com/p4lang/pna/pull/43#issuecomment-1293795491  I will update the PR to reflect that updated description.